File: sdk/java/src/main/kotlin/ConnectorClient.kt

/**
 * ConnectorClient — high-level wrapper around UniFFI bindings.
 *
 * Handles the full round-trip for any payment flow:
 *   1. Build connector HTTP request via {flow}_req_transformer (FFI)
 *   2. Execute the HTTP request via our standardized HttpClient
 *   3. Parse the connector response via {flow}_res_transformer (FFI)
 *
 * Flow methods (authorize, capture, void, refund, …) are defined as Kotlin
 * extension functions in GeneratedFlows.kt — no flow names are hardcoded here.
 * To add a new flow: edit sdk/flows.yaml and run `make generate`.
 *
 * Error Handling:
 *   FFI transformers return raw bytes that may represent either a success proto or
 *   an error proto (RequestError for req_transformer, ResponseError for res_transformer).
 *   On error, the decoded proto (RequestError or ResponseError) is thrown directly.
 *   Callers can catch the specific error type:
 *
 *       try {
 *           val response = client.authorize(request)
 *       } catch (e: RequestError) {
 *           println("${e.errorCode}: ${e.errorMessage}")
 *       } catch (e: ResponseError) {
 *           println("${e.errorCode}: ${e.errorMessage}")
 *       }
 */

package payments

import com.google.protobuf.ByteString
import com.google.protobuf.InvalidProtocolBufferException
import com.google.protobuf.MessageLite
import com.google.protobuf.Parser

open class ConnectorClient(
    val config: ConnectorConfig,
    val defaults: RequestConfig = RequestConfig.getDefaultInstance(),
    libPath: String? = null
) {
    private val httpClient: okhttp3.OkHttpClient

    init {
        // Instance-level cache: create the primary connection pool at startup
        val httpConfig = if (defaults.hasHttp()) defaults.http else null
        this.httpClient = HttpClient.createClient(httpConfig)
    }

    /**
     * Builds FfiOptions from config. Environment comes from ConnectorConfig (immutable).
     */
    private fun resolveFfiOptions(overrides: RequestConfig?): FfiOptions {
        return FfiOptions.newBuilder()
            .setEnvironment(config.environment)
            .setConnector(config.connector)
            .setAuth(config.auth)
            .build()
    }

    /**
     * Merges request-level HTTP overrides with client defaults.
     */
    private fun resolveHttpConfig(overrides: RequestConfig?): HttpConfig? {
        val clientHttp = if (defaults.hasHttp()) defaults.http else null
        val overrideHttp = if (overrides != null && overrides.hasHttp()) overrides.http else null

        if (overrideHttp == null) return clientHttp
        
        // Merge: Field-level override > Client default
        val builder = HttpConfig.newBuilder()
        if (clientHttp != null) {
            builder.mergeFrom(clientHttp)
        }
        builder.mergeFrom(overrideHttp)
        
        return builder.build()
    }


    /**
     * Parse FFI req_transformer bytes as either FfiConnectorHttpRequest or RequestError.
     * Tries the success type first; on failure parses as RequestError and throws
     * the RequestError directly.
     */
    private fun checkReqError(resultBytes: ByteArray): FfiConnectorHttpRequest {
        return try {
            FfiConnectorHttpRequest.parseFrom(resultBytes)
        } catch (e: InvalidProtocolBufferException) {
            // Try parsing as RequestError and throw directly
            throw RequestError.parseFrom(resultBytes)
        }
    }

    /**
     * Parse FFI res_transformer bytes as either success type or ResponseError.
     * Tries the success type first; on failure parses as ResponseError and throws
     * the ResponseError directly.
     */
    private fun <T : MessageLite> checkResError(
        resultBytes: ByteArray,
        successParser: Parser<T>
    ): T {
        return try {
            successParser.parseFrom(resultBytes)
        } catch (e: InvalidProtocolBufferException) {
            // Try parsing as ResponseError and throw directly
            throw ResponseError.parseFrom(resultBytes)
        }
    }


    /**
     * Execute a full round-trip for any payment flow.
     *
     * Errors from the FFI layer are thrown as RequestError or ResponseError directly.
     *
     * @param flow Flow name matching the FFI transformer prefix (e.g. "authorize").
     * @param requestBytes Serialized protobuf request bytes.
     * @param responseParser Protobuf parser for the expected response type.
     * @param options Optional RequestConfig message.
     * @return Parsed protobuf response.
     * @throws RequestError If the req_transformer returns an error proto.
     * @throws ResponseError If the res_transformer returns an error proto.
     */
    fun <T : MessageLite> executeFlow(
        flow: String,
        requestBytes: ByteArray,
        responseParser: Parser<T>,
        options: RequestConfig? = null,
    ): T {
        val reqTransformer = FlowRegistry.reqTransformers[flow]
            ?: error("Unknown flow: '$flow'")
        val resTransformer = FlowRegistry.resTransformers[flow]
            ?: error("Unknown flow: '$flow'")

        // 1. Resolve final configuration (Pattern-based merging)
        val ffiOptions = resolveFfiOptions(options)
        val optionsBytes = ffiOptions.toByteArray()
        val httpConfig = resolveHttpConfig(options)

        // 2. Build connector HTTP request via FFI
        //    Parse result bytes as FfiConnectorHttpRequest; if that fails, parse as RequestError.
        val connectorRequestBytes = reqTransformer(requestBytes, optionsBytes)
        val connectorRequest = checkReqError(connectorRequestBytes)

        val httpRequest = HttpRequest(
            url = connectorRequest.url,
            method = connectorRequest.method,
            headers = connectorRequest.headersMap,
            body = if (connectorRequest.hasBody()) connectorRequest.body.toByteArray() else null
        )

        // 3. Execute HTTP request via standardized HttpClient using the connection pool
        val response = HttpClient.execute(httpRequest, httpConfig, this.httpClient)

        // 4. Encode HTTP response as FfiConnectorHttpResponse protobuf bytes
        val ffiResponseBytes = FfiConnectorHttpResponse.newBuilder()
            .setStatusCode(response.statusCode)
            .putAllHeaders(response.headers)
            .setBody(ByteString.copyFrom(response.body))
            .build()
            .toByteArray()

        // 5. Parse connector response via FFI
        //    Parse result bytes as success type; if that fails, parse as ResponseError.
        val resultBytes = resTransformer(
            ffiResponseBytes,
            requestBytes,
            optionsBytes,
        )

        return checkResError(resultBytes, responseParser)
    }


    /**
     * Execute a single-step flow directly via FFI (no HTTP round-trip).
     * Used for inbound flows like webhook processing where the connector sends data to us.
     *
     * Errors are thrown as ResponseError directly.
     *
     * @param flow Flow name matching the FFI transformer (e.g. "handle").
     * @param requestBytes Serialized protobuf request bytes.
     * @param responseParser Protobuf parser for the expected response type.
     * @param options Optional RequestConfig for FFI context. Merged with client defaults.
     * @return Parsed protobuf response.
     * @throws ResponseError If the FFI transformer returns an error proto.
     */
    fun <T : MessageLite> executeDirect(
        flow: String,
        requestBytes: ByteArray,
        responseParser: Parser<T>,
        options: RequestConfig? = null,
    ): T {
        val transformer = FlowRegistry.directTransformers[flow]
            ?: error("Unknown single-step flow: '$flow'. Register it via a {flow}_transformer in services/payments.rs and run `make generate`.")

        val ffiOptions = resolveFfiOptions(options)
        val optionsBytes = ffiOptions.toByteArray()

        val resultBytes = transformer(requestBytes, optionsBytes)
        
        // Parse result bytes as success type; if that fails, parse as ResponseError.
        return checkResError(resultBytes, responseParser)
    }
}

File: sdk/rust/src/connector_client.rs
use std::collections::HashMap;
use std::error::Error;

use crate::http_client::{
    HttpClient, HttpClientError, HttpOptions as NativeHttpOptions, HttpRequest as ClientHttpRequest,
};
use connector_service_ffi::handlers::payments::{authorize_req_handler, authorize_res_handler};
use connector_service_ffi::types::{FfiMetadataPayload, FfiRequestData};
use connector_service_ffi::utils::ffi_headers_to_masked_metadata;
use domain_types::router_data::ConnectorSpecificConfig;
use domain_types::router_response_types::Response;
use domain_types::utils::ForeignTryFrom;
use grpc_api_types::payments::{
    ConnectorConfig, FfiOptions, PaymentServiceAuthorizeRequest, PaymentServiceAuthorizeResponse,
    RequestConfig,
};

/// ConnectorClient — high-level Rust wrapper for the Connector Service.
///
/// Handles the full round-trip for any payment flow:
///   1. Build connector HTTP request via Rust core handlers
///   2. Execute the HTTP request via our standardized HttpClient (reqwest)
///   3. Parse the connector response via Rust core handlers
///
/// This client owns its primary connection pool (http_client).
pub struct ConnectorClient {
    http_client: HttpClient,
    config: ConnectorConfig,
    #[allow(dead_code)] // reserved for merging with per-request RequestConfig (e.g. http)
    defaults: RequestConfig,
}

impl ConnectorClient {
    /// Initialize a new ConnectorClient.
    ///
    /// # Arguments
    /// * `config` - The ConnectorConfig (connector_config with typed auth, options with environment).
    /// * `options` - Optional RequestConfig for default http/vault settings.
    pub fn new(
        config: ConnectorConfig,
        options: Option<RequestConfig>,
    ) -> Result<Self, HttpClientError> {
        let defaults = options.unwrap_or_default();

        // Map the Protobuf options to native transport options
        let native_opts = match defaults.http.as_ref() {
            Some(http_proto) => NativeHttpOptions::from(http_proto),
            None => NativeHttpOptions::default(),
        };

        let http_client = HttpClient::new(native_opts)?;

        Ok(Self {
            http_client,
            config,
            defaults,
        })
    }

    /// Builds FfiOptions from config. Environment comes from SdkOptions (immutable).
    fn resolve_ffi_options(&self, _options: &Option<RequestConfig>) -> FfiOptions {
        let environment = self
            .config
            .options
            .as_ref()
            .map(|o| o.environment)
            .unwrap_or(0);
        FfiOptions {
            environment,
            connector_config: self.config.connector_config.clone(),
        }
    }

    /// Authorize a payment flow.
    ///
    /// # Arguments
    /// * `request` - The PaymentServiceAuthorizeRequest protobuf message.
    /// * `metadata` - Metadata map containing x-* headers for MaskedMetadata.
    /// * `options` - Optional RequestConfig for per-call overrides (http, vault).
    pub async fn authorize(
        &self,
        request: PaymentServiceAuthorizeRequest,
        metadata: &HashMap<String, String>,
        options: Option<RequestConfig>,
    ) -> Result<PaymentServiceAuthorizeResponse, Box<dyn Error>> {
        // 1. Resolve final configuration
        let ffi_options = self.resolve_ffi_options(&options);

        let override_opts = options
            .as_ref()
            .and_then(|o| o.http.as_ref())
            .map(NativeHttpOptions::from);

        let ffi_request = build_ffi_request(request.clone(), metadata, &ffi_options)?;
        let environment = Some(grpc_api_types::payments::Environment::try_from(
            ffi_options.environment,
        )?);

        // 2. Build the connector HTTP request via core handler
        let connector_request = authorize_req_handler(ffi_request, environment)
            .map_err(|e| format!("authorize_req_handler failed: {:?}", e))?
            .ok_or("No connector request generated")?;

        // 3. Execute HTTP using the instance-owned client and potential overrides
        let (body, boundary) = connector_request
            .body
            .as_ref()
            .map(|b| b.get_body_bytes())
            .transpose()
            .map_err(|e| format!("Body extraction failed: {e}"))?
            .unwrap_or((None, None));
        let mut headers = connector_request.get_headers_map();

        if let Some(boundary) = boundary {
            headers.insert(
                "content-type".to_string(),
                format!("multipart/form-data; boundary={}", boundary),
            );
        }

        let http_req = ClientHttpRequest {
            url: connector_request.url.clone(),
            method: connector_request.method,
            headers,
            body,
        };

        let http_response = self.http_client.execute(http_req, override_opts).await?;

        // 4. Convert HTTP response to domain Response type
        let mut header_map = http::HeaderMap::new();
        for (key, value) in &http_response.headers {
            if let Ok(name) = http::header::HeaderName::from_bytes(key.as_bytes()) {
                if let Ok(val) = http::header::HeaderValue::from_bytes(value.as_bytes()) {
                    header_map.insert(name, val);
                }
            }
        }

        let response = Response {
            headers: Some(header_map),
            response: bytes::Bytes::from(http_response.body),
            status_code: http_response.status_code,
        };

        // 5. Parse response via core handler
        let ffi_request_for_res = build_ffi_request(request, metadata, &ffi_options)?;
        match authorize_res_handler(ffi_request_for_res, response, environment) {
            Ok(auth_response) => Ok(auth_response),
            Err(error_response) => {
                Err(format!("Authorization failed: {:?}", error_response).into())
            }
        }
    }
}

/// Internal helper to build the context-heavy FfiRequestData from raw inputs.
pub fn build_ffi_request<T>(
    payload: T,
    metadata: &HashMap<String, String>,
    options: &FfiOptions,
) -> Result<FfiRequestData<T>, Box<dyn Error>> {
    let proto_config = options
        .connector_config
        .as_ref()
        .ok_or("Missing connector_config in FfiOptions")?;

    let config_variant = proto_config
        .config
        .as_ref()
        .ok_or("Missing config variant in ConnectorSpecificConfig")?;

    let connector =
        domain_types::connector_types::ConnectorEnum::foreign_try_from(config_variant.clone())
            .map_err(|e| format!("Connector mapping failed: {e}"))?;

    let connector_config = ConnectorSpecificConfig::foreign_try_from(proto_config.clone())
        .map_err(|e| format!("Connector config mapping failed: {e}"))?;

    let masked_metadata = ffi_headers_to_masked_metadata(metadata)
        .map_err(|e| format!("Metadata mapping failed: {:?}", e))?;

    Ok(FfiRequestData {
        payload,
        extracted_metadata: FfiMetadataPayload {
            connector,
            connector_config,
        },
        masked_metadata: Some(masked_metadata),
    })
}
